In [1]:
# Dependencies
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as st
import numpy as np

In [2]:
# Read in file
happiness_df = pd.read_csv('world-happiness-report-2015-2022-cleaned.csv')
happiness_df

Unnamed: 0  Happiness Rank      Country  \
0              0               1  Switzerland   
1              1               2      Iceland   
2              2               3      Denmark   
3              3               4       Norway   
4              4               5       Canada   
...          ...             ...          ...   
1224         141             142    Botswana*   
1225         142             143      Rwanda*   
1226         143             144     Zimbabwe   
1227         144             145      Lebanon   
1228         145             146  Afghanistan   

                               Region Happiness Score  \
0                      Western Europe           7.587   
1                      Western Europe           7.561   
2                      Western Europe           7.527   
3                      Western Europe           7.522   
4                       North America           7.427   
...                               ...             ...   
1224                                -           3,471   
1225                                -           3,268   
1226               Sub-Saharan Africa           2,995   
1227  Middle East and Northern Africa           2,955   
1228                    Southern Asia           2,404   

     Economy (GDP per Capita) Family (Social Support)  \
0                     1.39651                 1.34951   
1                     1.30232                 1.40223   
2                     1.32548                 1.36058   
3                       1.459                 1.33095   
4                     1.32629                 1.32261   
...                       ...                     ...   
1224                    1,503                   0,815   
1225                    0,785                   0,133   
1226                    0,947                   0,690   
1227                    1,392                   0,498   
1228                    0,758                   0,000   

     Health (Life Expectancy)  Freedom Trust (Government Corruption)  \
0                     0.94143  0.66557                       0.41978   
1                     0.94784  0.62877                       0.14145   
2                     0.87464  0.64938                       0.48357   
3                     0.88521  0.66973                       0.36503   
4                     0.90563  0.63297                       0.32957   
...                       ...      ...                           ...   
1224                    0,280    0,571                         0,102   
1225                    0,462    0,621                         0,544   
1226                    0,270    0,329                         0,105   
1227                    0,631    0,103                         0,034   
1228                    0,289    0,000                         0,005   

     Generosity  Year  
0       0.29678  2015  
1        0.4363  2015  
2       0.34139  2015  
3       0.34699  2015  
4       0.45811  2015  
...         ...   ...  
1224      0,012  2022  
1225      0,187  2022  
1226      0,106  2022  
1227      0,082  2022  
1228      0,089  2022  

[1229 rows x 12 columns]

In [3]:
# Removing all regions marked with no value "-"  -Tye
happiness_df.drop(happiness_df[happiness_df['Region'] == "-"].index, inplace = True)

In [4]:
#Converting Happiness Score to Float as some data contains (,) versus (.) as a decimal.-Tye
happiness_df["Happiness Score"] = happiness_df["Happiness Score"].str.replace(',','.').astype(float)

In [5]:
happiness_df.dtypes

Unnamed: 0                         int64
Happiness Rank                     int64
Country                           object
Region                            object
Happiness Score                  float64
Economy (GDP per Capita)          object
Family (Social Support)           object
Health (Life Expectancy)          object
Freedom                           object
Trust (Government Corruption)     object
Generosity                        object
Year                               int64
dtype: object

In [6]:
#complete data cleaning steps 

happiness_df.count()

Unnamed: 0                       1183
Happiness Rank                   1183
Country                          1183
Region                           1183
Happiness Score                  1183
Economy (GDP per Capita)         1183
Family (Social Support)          1183
Health (Life Expectancy)         1183
Freedom                          1183
Trust (Government Corruption)    1183
Generosity                       1183
Year                             1183
dtype: int64

In [8]:
#Unique Number of countries in our data -Tye
total_countries = len(happiness_df['Country'].unique())
print(f'Number of total countries in the data set: {total_countries}')

Number of total countries in the data set: 169


In [9]:
#Unique Number of countries in our data -Tye
total_Regions = len(happiness_df['Region'].unique())
print(f'Number of total countries in the data set: {total_Regions}')

Number of total countries in the data set: 16


In [10]:
#Unique Values by column -Tye
happiness_df.nunique()


Unnamed: 0                        158
Happiness Rank                    158
Country                           169
Region                             16
Happiness Score                  1062
Economy (GDP per Capita)         1115
Family (Social Support)          1092
Health (Life Expectancy)         1040
Freedom                          1018
Trust (Government Corruption)     923
Generosity                        951
Year                                8
dtype: int64

**Understanding the Data:**
Our date consists of 195 countries, 17 regions, and over the course of 8 years. We need to be careful with this data as we can see that all factors being tested for happiness are not equal. This means most likely some countries were not able to be evaluated by one of the variables of happiness, or as we have seen in our data that 2015 data was stored using "." in the data, while more recent years (2022) use "," to configure theirt data. This returns a higher unique number in these columns.
-Tye

In [11]:
#Region Names -Tye
Region_Names= happiness_df.Region.unique()
print(Region_Names)

['Western Europe' 'North America' 'Australia and New Zealand'
 'Middle East and Northern Africa' 'Latin America and Caribbean'
 'Southeastern Asia' 'Central and Eastern Europe' 'Eastern Asia'
 'Sub-Saharan Africa' 'Southern Asia' 'North America and ANZ'
 'Middle East and North Africa' 'East Asia' 'Southeast Asia'
 'Commonwealth of Independent States' 'South Asia']


In [12]:
Region = happiness_df.set_index(["Region"]).groupby(["Region"])
Region

In [13]:
# Data Frame created to analyse year over year Happiness Rank by Region
rank_2015 = happiness_df[happiness_df['Year'] == 2015].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2016 = happiness_df[happiness_df["Year"] == 2016].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2017 = happiness_df[happiness_df["Year"] == 2017].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2018 = happiness_df[happiness_df["Year"] == 2018].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2019= happiness_df[happiness_df["Year"] == 2019].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2020= happiness_df[happiness_df["Year"] == 2020].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2021= happiness_df[happiness_df["Year"] == 2021].groupby(happiness_df.Region).mean()["Happiness Rank"]
rank_2022= happiness_df[happiness_df["Year"] == 2022].groupby(happiness_df.Region).mean()["Happiness Rank"]
Region_by_year_mean = pd.DataFrame({"2015":rank_2015, "2016":rank_2016, 
                                     "2017":rank_2017, "2018":rank_2018, "2019":rank_2019, "2020":rank_2020,"2021":rank_2021,"2022":rank_2022})
Region_by_year_mean.index.name = "Region"


Region_by_year_mean[["2015","2016","2017","2018","2019","2020","2021","2022"]] = Region_by_year_mean[["2015","2016","2017","2018","2019","2020","2021","2022"]].applymap("{:.2f}".format)

Region_by_year_mean

C:\Users\laphe\AppData\Local\Temp\ipykernel_21048\556511397.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  rank_2015 = happiness_df[happiness_df['Year'] == 2015].groupby(happiness_df.Region).mean()["Happiness Rank"]
C:\Users\laphe\AppData\Local\Temp\ipykernel_21048\556511397.py:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  rank_2016 = happiness_df[happiness_df["Year"] == 2016].groupby(happiness_df.Region).mean()["Happiness Rank"]
C:\Users\laphe\AppData\Local\Temp\ipykernel_21048\556511397.py:4: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future versi

2015    2016    2017    2018    2019  \
Region                                                                       
Australia and New Zealand             9.50    8.50    9.00    9.00    9.50   
Central and Eastern Europe           79.00   78.45   75.07   74.97   70.50   
Commonwealth of Independent States     nan     nan     nan     nan     nan   
East Asia                              nan     nan     nan     nan     nan   
Eastern Asia                         64.50   67.17   71.25   65.50   64.83   
Latin America and Caribbean          46.91   48.33   50.81   51.25   52.00   
Middle East and North Africa           nan     nan     nan     nan     nan   
Middle East and Northern Africa      77.60   78.11   76.42   84.28   84.89   
North America                        10.00    9.50   10.50   12.50   14.00   
North America and ANZ                  nan     nan     nan     nan     nan   
South Asia                             nan     nan     nan     nan     nan   
Southeast Asia                         nan     nan     nan     nan     nan   
Southeastern Asia                    81.22   80.00   73.75   81.89   85.11   
Southern Asia                       113.14  111.71  109.86  111.71  115.86   
Sub-Saharan Africa                  127.90  129.66  128.78  127.19  124.75   
Western Europe                       29.52   29.19   27.14   23.00   21.35   

                                      2020    2021    2022  
Region                                                      
Australia and New Zealand              nan     nan   11.00  
Central and Eastern Europe           58.29   53.24   58.42  
Commonwealth of Independent States   82.58   78.25     nan  
East Asia                            66.83   62.17     nan  
Eastern Asia                           nan     nan   63.25  
Latin America and Caribbean          52.38   56.65   56.89  
Middle East and North Africa         89.12   90.00     nan  
Middle East and Northern Africa        nan     nan   87.08  
North America                          nan     nan   15.50  
North America and ANZ                12.25   13.25     nan  
South Asia                          111.29  114.14     nan  
Southeast Asia                       81.00   81.00     nan  
Southeastern Asia                      nan     nan   79.67  
Southern Asia                          nan     nan  118.00  
Sub-Saharan Africa                  121.87  118.47  115.09  
Western Europe                       22.52   21.38   19.58

**Note:** Year over year mean rank by region (2015-2022)
Western Europe & Australia/New Zealand consistantly rank among the happiest regions in the world through the data time (2015-2022). Follwoing the beginning of the Covid-19 Pandemic Western Europe has consistantly been the happiest region, as Southern Asia is amongst the worst with a median 118.0 rank out of 169 countries evaluated. -Tye

In [14]:
# DataFrame Analysing the median Happiness Score by world region -Tye
score_2015 = happiness_df[happiness_df['Year'] == 2015].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2016 = happiness_df[happiness_df["Year"] == 2016].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2017 = happiness_df[happiness_df["Year"] == 2017].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2018 = happiness_df[happiness_df["Year"] == 2018].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2019= happiness_df[happiness_df["Year"] == 2019].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2020= happiness_df[happiness_df["Year"] == 2020].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2021= happiness_df[happiness_df["Year"] == 2021].groupby(happiness_df.Region).mean()["Happiness Score"]
score_2022= happiness_df[happiness_df["Year"] == 2022].groupby(happiness_df.Region).mean()["Happiness Score"]
region_score_by_year_mean = pd.DataFrame({"2015":score_2015, "2016":score_2016, 
                                     "2017":score_2017, "2018":score_2018, "2019":score_2019, "2020":score_2020,"2021":score_2021,"2022":score_2022})
region_score_by_year_mean.index.name = "Region"


region_score_by_year_mean[["2015","2016","2017","2018","2019","2020","2021","2022"]] = region_score_by_year_mean[["2015","2016","2017","2018","2019","2020","2021","2022"]].applymap("{:.2f}".format)

region_score_by_year_mean

C:\Users\laphe\AppData\Local\Temp\ipykernel_21048\4050093992.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  score_2015 = happiness_df[happiness_df['Year'] == 2015].groupby(happiness_df.Region).mean()["Happiness Score"]
C:\Users\laphe\AppData\Local\Temp\ipykernel_21048\4050093992.py:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  score_2016 = happiness_df[happiness_df["Year"] == 2016].groupby(happiness_df.Region).mean()["Happiness Score"]
C:\Users\laphe\AppData\Local\Temp\ipykernel_21048\4050093992.py:4: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a futur

2015  2016  2017  2018  2019  2020  2021  \
Region                                                                         
Australia and New Zealand           7.29  7.32  7.30  7.30  7.27   nan   nan   
Central and Eastern Europe          5.33  5.37  5.41  5.46  5.57  5.88  5.98   
Commonwealth of Independent States   nan   nan   nan   nan   nan  5.36  5.47   
East Asia                            nan   nan   nan   nan   nan  5.71  5.81   
Eastern Asia                        5.63  5.62  5.50  5.67  5.69   nan   nan   
Latin America and Caribbean         6.14  6.10  5.96  5.94  5.94  5.98  5.91   
Middle East and North Africa         nan   nan   nan   nan   nan  5.23  5.22   
Middle East and Northern Africa     5.41  5.39  5.37  5.20  5.24   nan   nan   
North America                       7.27  7.25  7.15  7.11  7.08   nan   nan   
North America and ANZ                nan   nan   nan   nan   nan  7.17  7.13   
South Asia                           nan   nan   nan   nan   nan  4.48  4.44   
Southeast Asia                       nan   nan   nan   nan   nan  5.38  5.41   
Southeastern Asia                   5.32  5.34  5.44  5.31  5.27   nan   nan   
Southern Asia                       4.58  4.56  4.63  4.60  4.53   nan   nan   
Sub-Saharan Africa                  4.20  4.14  4.08  4.19  4.31  4.38  4.49   
Western Europe                      6.69  6.69  6.70  6.83  6.90  6.90  6.91   

                                    2022  
Region                                    
Australia and New Zealand           7.18  
Central and Eastern Europe          5.90  
Commonwealth of Independent States   nan  
East Asia                            nan  
Eastern Asia                        5.83  
Latin America and Caribbean         5.93  
Middle East and North Africa         nan  
Middle East and Northern Africa     5.21  
North America                       7.00  
North America and ANZ                nan  
South Asia                           nan  
Southeast Asia                       nan  
Southeastern Asia                   5.43  
Southern Asia                       4.27  
Sub-Saharan Africa                  4.54  
Western Europe                      6.94

The Analysis of Happiness Score shows Australia/New Zealand and Western Europe not only with the highest score, but one of the most consistant. The happiness score of North America has declined year over year since 2015, with it being the 2nd highest world region in 2022. - Tye

In [15]:
#Statistical Analysis of World Region by Happiness Rank -Tye
region_statistics_df = happiness_df.groupby('Region')
Mean = region_statistics_df['Happiness Rank'].mean()
Median = region_statistics_df['Happiness Rank'].median()
Variance = region_statistics_df['Happiness Rank'].var()
Standard_Deviation = region_statistics_df['Happiness Rank'].std()
Standard_Error_Of_Mean = region_statistics_df['Happiness Rank'].sem()

summary_statistics_df = pd.DataFrame({
    'Mean': Mean,
    'Median': Median,
    'Variance': Variance,
    'Standard deviation': Standard_Deviation,
    'SEM': Standard_Error_Of_Mean
})
# Present Summary Statistics
print('Method I')
summary_statistics_df

Method I


Mean  Median     Variance  \
Region                                                                
Australia and New Zealand             9.416667     9.5     1.719697   
Central and Eastern Europe           70.242574    67.5   737.637382   
Commonwealth of Independent States   80.416667    75.5   564.688406   
East Asia                            64.500000    66.0   471.727273   
Eastern Asia                         65.937500    63.5   449.995968   
Latin America and Caribbean          51.692771    45.0   830.008069   
Middle East and North Africa         89.558824   107.5  1869.223708   
Middle East and Northern Africa      81.018519    86.5  1759.644514   
North America                        12.000000    13.5    24.363636   
North America and ANZ                12.750000    11.5    15.928571   
South Asia                          112.714286   106.0   761.450549   
Southeast Asia                       81.000000    82.0   879.529412   
Southeastern Asia                    80.396226    81.0  1178.359216   
Southern Asia                       113.268293   116.0   539.001220   
Sub-Saharan Africa                  124.731449   128.0   482.225497   
Western Europe                       24.292683    16.5   597.202155   

                                    Standard deviation       SEM  
Region                                                            
Australia and New Zealand                     1.311372  0.378561  
Central and Eastern Europe                   27.159481  1.910934  
Commonwealth of Independent States           23.763173  4.850637  
East Asia                                    21.719283  6.269817  
Eastern Asia                                 21.213108  3.749983  
Latin America and Caribbean                  28.809861  2.236079  
Middle East and North Africa                 43.234520  7.414659  
Middle East and Northern Africa              41.948117  4.036459  
North America                                 4.935953  1.424887  
North America and ANZ                         3.991061  1.411053  
South Asia                                   27.594393  7.374912  
Southeast Asia                               29.656861  6.990189  
Southeastern Asia                            34.327237  4.715209  
Southern Asia                                23.216400  3.625792  
Sub-Saharan Africa                           21.959633  1.305365  
Western Europe                               24.437720  1.908265

Statistical Analysis shows the happiest region in the world is Australia/New Zealand across every statistical measure, while Sub-Saharan Africa are one of the least happy by rank. -Tye

In [16]:
#-Lauren
#happiness_df.sort_values('Happiness Rank').head(8)

In [73]:
#-Lauren
#out = happiness_df.groupby(pd.to_datetime(happiness_df['Year']).dt.strftime('%m/%d/%Y'))['Country'].agg(['first', 'last']).reset_index()
#out.head(5)

#df2 = (happiness_df.groupby(pd.to_datetime(happiness_df['Year']).dt.date)
         #['Country'].agg(['first', 'last'])
       #)
#df2.head(5)

#mins = happiness_df.groupby('Year')['Happiness Rank'].min()
#maxs = happiness_df.groupby('Region')['Happiness Rank'].max()
#df2=pd.concat([mins,maxs],axis=1)
#df2.columns=(['Min Happiness Rank','Max Happiness Rank'])
#df2['Year']=mins.index
#df2.head(8)

#happiness_df.groupby(['Year'])['Happiness Rank'].agg(['min', 'max'])


min  max
Year          
2015    1  158
2016    1  157
2017    1  155
2018    1  156
2019    1  155
2020    1  153
2021    1  149
2022    1  146

In [17]:
#Looking at dataframe -Lauren
happiness_df

Unnamed: 0  Happiness Rank       Country  \
0              0               1   Switzerland   
1              1               2       Iceland   
2              2               3       Denmark   
3              3               4        Norway   
4              4               5        Canada   
...          ...             ...           ...   
1221         138             139      Tanzania   
1222         139             140  Sierra Leone   
1226         143             144      Zimbabwe   
1227         144             145       Lebanon   
1228         145             146   Afghanistan   

                               Region  Happiness Score  \
0                      Western Europe            7.587   
1                      Western Europe            7.561   
2                      Western Europe            7.527   
3                      Western Europe            7.522   
4                       North America            7.427   
...                               ...              ...   
1221               Sub-Saharan Africa            3.702   
1222               Sub-Saharan Africa            3.574   
1226               Sub-Saharan Africa            2.995   
1227  Middle East and Northern Africa            2.955   
1228                    Southern Asia            2.404   

     Economy (GDP per Capita) Family (Social Support)  \
0                     1.39651                 1.34951   
1                     1.30232                 1.40223   
2                     1.32548                 1.36058   
3                       1.459                 1.33095   
4                     1.32629                 1.32261   
...                       ...                     ...   
1221                    0,848                   0,597   
1222                    0,686                   0,416   
1226                    0,947                   0,690   
1227                    1,392                   0,498   
1228                    0,758                   0,000   

     Health (Life Expectancy)  Freedom Trust (Government Corruption)  \
0                     0.94143  0.66557                       0.41978   
1                     0.94784  0.62877                       0.14145   
2                     0.87464  0.64938                       0.48357   
3                     0.88521  0.66973                       0.36503   
4                     0.90563  0.63297                       0.32957   
...                       ...      ...                           ...   
1221                    0,425    0,578                         0,270   
1222                    0,273    0,387                         0,055   
1226                    0,270    0,329                         0,105   
1227                    0,631    0,103                         0,034   
1228                    0,289    0,000                         0,005   

     Generosity  Year  
0       0.29678  2015  
1        0.4363  2015  
2       0.34139  2015  
3       0.34699  2015  
4       0.45811  2015  
...         ...   ...  
1221      0,248  2022  
1222      0,202  2022  
1226      0,106  2022  
1227      0,082  2022  
1228      0,089  2022  

[1183 rows x 12 columns]

In [18]:
#Make a copy of happiness_df
reduced_happiness_df = happiness_df.copy()
reduced_happiness_df

Unnamed: 0  Happiness Rank       Country  \
0              0               1   Switzerland   
1              1               2       Iceland   
2              2               3       Denmark   
3              3               4        Norway   
4              4               5        Canada   
...          ...             ...           ...   
1221         138             139      Tanzania   
1222         139             140  Sierra Leone   
1226         143             144      Zimbabwe   
1227         144             145       Lebanon   
1228         145             146   Afghanistan   

                               Region  Happiness Score  \
0                      Western Europe            7.587   
1                      Western Europe            7.561   
2                      Western Europe            7.527   
3                      Western Europe            7.522   
4                       North America            7.427   
...                               ...              ...   
1221               Sub-Saharan Africa            3.702   
1222               Sub-Saharan Africa            3.574   
1226               Sub-Saharan Africa            2.995   
1227  Middle East and Northern Africa            2.955   
1228                    Southern Asia            2.404   

     Economy (GDP per Capita) Family (Social Support)  \
0                     1.39651                 1.34951   
1                     1.30232                 1.40223   
2                     1.32548                 1.36058   
3                       1.459                 1.33095   
4                     1.32629                 1.32261   
...                       ...                     ...   
1221                    0,848                   0,597   
1222                    0,686                   0,416   
1226                    0,947                   0,690   
1227                    1,392                   0,498   
1228                    0,758                   0,000   

     Health (Life Expectancy)  Freedom Trust (Government Corruption)  \
0                     0.94143  0.66557                       0.41978   
1                     0.94784  0.62877                       0.14145   
2                     0.87464  0.64938                       0.48357   
3                     0.88521  0.66973                       0.36503   
4                     0.90563  0.63297                       0.32957   
...                       ...      ...                           ...   
1221                    0,425    0,578                         0,270   
1222                    0,273    0,387                         0,055   
1226                    0,270    0,329                         0,105   
1227                    0,631    0,103                         0,034   
1228                    0,289    0,000                         0,005   

     Generosity  Year  
0       0.29678  2015  
1        0.4363  2015  
2       0.34139  2015  
3       0.34699  2015  
4       0.45811  2015  
...         ...   ...  
1221      0,248  2022  
1222      0,202  2022  
1226      0,106  2022  
1227      0,082  2022  
1228      0,089  2022  

[1183 rows x 12 columns]

In [19]:
#Reduce dataframe to Country, Happiness Score and Year
reduced_happiness_df = reduced_happiness_df[["Country", "Happiness Score", "Year"]]
reduced_happiness_df

Country  Happiness Score  Year
0      Switzerland            7.587  2015
1          Iceland            7.561  2015
2          Denmark            7.527  2015
3           Norway            7.522  2015
4           Canada            7.427  2015
...            ...              ...   ...
1221      Tanzania            3.702  2022
1222  Sierra Leone            3.574  2022
1226      Zimbabwe            2.995  2022
1227       Lebanon            2.955  2022
1228   Afghanistan            2.404  2022

[1183 rows x 3 columns]

In [34]:
top_2015 = reduced_happiness_df[reduced_happiness_df['Year'] == 2015].groupby(reduced_happiness_df.Country).max()["Happiness Score"]
top_2015_df = top_2015.sort_values(ascending=False)

top_2015_df


Country
Switzerland    7.587
Iceland        7.561
Denmark        7.527
Norway         7.522
Canada         7.427
               ...  
Rwanda         3.465
Benin          3.340
Syria          3.006
Burundi        2.905
Togo           2.839
Name: Happiness Score, Length: 158, dtype: float64

In [36]:
# Extract first and last value of dataset
df2015 = top_2015.iloc[[0, -1]]
df2015

Country
Afghanistan    3.575
Zimbabwe       4.610
Name: Happiness Score, dtype: float64

In [37]:
top_2016 = happiness_df[happiness_df['Year'] == 2016].groupby(happiness_df.Country).max()["Happiness Score"]
top_2016.df = top_2016.sort_values(ascending=False)

top_2016.df


Country
Denmark        7.526
Switzerland    7.509
Iceland        7.501
Norway         7.498
Finland        7.413
               ...  
Benin          3.484
Afghanistan    3.360
Togo           3.303
Syria          3.069
Burundi        2.905
Name: Happiness Score, Length: 157, dtype: float64

In [38]:
df2016 = top_2016.df.iloc[[0, -1]]
df2016

Country
Denmark    7.526
Burundi    2.905
Name: Happiness Score, dtype: float64

In [39]:
top_2017 = happiness_df[happiness_df['Year'] == 2017].groupby(happiness_df.Country).max()["Happiness Score"]
top_2017.df = top_2017.sort_values(ascending=False)

top_2017.df


Country
Norway                      7.537
Denmark                     7.522
Iceland                     7.504
Switzerland                 7.494
Finland                     7.469
                            ...  
Rwanda                      3.471
Syria                       3.462
Tanzania                    3.349
Burundi                     2.905
Central African Republic    2.693
Name: Happiness Score, Length: 149, dtype: float64

In [40]:
df2017 = top_2017.df.iloc[[0, -1]]
df2017

Country
Norway                      7.537
Central African Republic    2.693
Name: Happiness Score, dtype: float64

In [41]:
top_2018 = happiness_df[happiness_df['Year'] == 2018].groupby(happiness_df.Country).max()["Happiness Score"]
top_2018.df = top_2018.sort_values(ascending=False)

top_2018.df


Country
Finland                     7.632
Norway                      7.594
Denmark                     7.555
Iceland                     7.495
Switzerland                 7.487
                            ...  
Rwanda                      3.408
Yemen                       3.355
Tanzania                    3.303
Central African Republic    3.083
Burundi                     2.905
Name: Happiness Score, Length: 149, dtype: float64

In [42]:
df2018 = top_2018.df.iloc[[0, -1]]
df2018

Country
Finland    7.632
Burundi    2.905
Name: Happiness Score, dtype: float64

In [43]:
top_2019 = happiness_df[happiness_df['Year'] == 2019].groupby(happiness_df.Country).max()["Happiness Score"]
top_2019.df = top_2019.sort_values(ascending=False)

top_2019.df


Country
Finland                     7.769
Denmark                     7.600
Norway                      7.554
Iceland                     7.494
Netherlands                 7.488
                            ...  
Yemen                       3.380
Rwanda                      3.334
Tanzania                    3.231
Afghanistan                 3.203
Central African Republic    3.083
Name: Happiness Score, Length: 149, dtype: float64

In [44]:
df2019 = top_2019.df.iloc[[0, -1]]
df2019

Country
Finland                     7.769
Central African Republic    3.083
Name: Happiness Score, dtype: float64

In [45]:
top_2020 = happiness_df[happiness_df['Year'] == 2020].groupby(happiness_df.Country).max()["Happiness Score"]
top_2020.df = top_2020.sort_values(ascending=False)

top_2020.df


Country
Finland                     7.8087
Denmark                     7.6456
Switzerland                 7.5599
Iceland                     7.5045
Norway                      7.4880
                             ...  
Central African Republic    3.4759
Rwanda                      3.3123
Zimbabwe                    3.2992
South Sudan                 2.8166
Afghanistan                 2.5669
Name: Happiness Score, Length: 153, dtype: float64

In [46]:
df2020 = top_2020.df.iloc[[0, -1]]
df2020

Country
Finland        7.8087
Afghanistan    2.5669
Name: Happiness Score, dtype: float64

In [47]:
top_2021 = happiness_df[happiness_df['Year'] == 2021].groupby(happiness_df.Country).max()["Happiness Score"]
top_2021.df = top_2021.sort_values(ascending=False)

top_2021.df


Country
Finland        7.842
Denmark        7.620
Switzerland    7.571
Iceland        7.554
Netherlands    7.464
               ...  
Lesotho        3.512
Botswana       3.467
Rwanda         3.415
Zimbabwe       3.145
Afghanistan    2.523
Name: Happiness Score, Length: 149, dtype: float64

In [48]:
df2021 = top_2021.df.iloc[[0, -1]]
df2021

Country
Finland        7.842
Afghanistan    2.523
Name: Happiness Score, dtype: float64

In [49]:
top_2022 = happiness_df[happiness_df['Year'] == 2022].groupby(happiness_df.Country).max()["Happiness Score"]
top_2022.df = top_2022.sort_values(ascending=False)

top_2022.df


Country
Finland         7.821
Denmark         7.636
Iceland         7.557
Switzerland     7.512
Netherlands     7.415
                ...  
Tanzania        3.702
Sierra Leone    3.574
Zimbabwe        2.995
Lebanon         2.955
Afghanistan     2.404
Name: Happiness Score, Length: 119, dtype: float64

In [50]:
df2022 = top_2022.df.iloc[[0, -1]]
df2022

Country
Finland        7.821
Afghanistan    2.404
Name: Happiness Score, dtype: float64

In [51]:
from functools import reduce


In [53]:
#define list of DataFrames
dfs = [df2015, df2016, df2017, df2018, df2019, df2020, df2021, df2022]

#merge all DataFrames into one
final_df = reduce(lambda  left,right: pd.merge(left,right,on=['Country', 'Happiness Score'],
                                            how='outer'), dfs)
final_df

Happiness Score
Country                                  
Afghanistan                        3.5750
Zimbabwe                           4.6100
Denmark                            7.5260
Burundi                            2.9050
Norway                             7.5370
Central African Republic           2.6930
Finland                            7.6320
Finland                            7.7690
Central African Republic           3.0830
Finland                            7.8087
Afghanistan                        2.5669
Finland                            7.8420
Afghanistan                        2.5230
Finland                            7.8210
Afghanistan                        2.4040

In [133]:
#final_df
#x_axis = final_df["Country"]
#tumsiz = final_df["Happiness Score"]

#fig1, ax1 = plt.subplots()
#plt.title('Highest and Lowest Happiness Scores (2015-2022)')
#plt.plot(x_axis, tumsiz,linewidth=2, markersize=15,marker="o",color="blue")
#plt.xlabel('Country')
#plt.ylabel('Happiness Score')

#Skeleton...
# Plot the world average as a line chart
world_avg, = plt.plot(years, average_unemployment, color="blue", label="World Average" )

# Plot the unemployment values for a single country
country_one, = plt.plot(years, combined_unemployed_data.loc['USA',["2010","2011","2012","2013","2014","2015",
                                                                  "2016","2017","2018","2019","2020"]], 
                        color="green",label=combined_unemployed_data.loc['USA',"Country Name"])

# Create a legend for our chart
plt.legend(handles=[world_avg, country_one], loc="best")

# Show the chart
plt.show()

KeyError: 'Country'